src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ArenaAllocator.java

```diff
@@ -51,99 +60,48 @@ MemorySegment trySlice(long bytesSize, long bytesAlignment) {
         }
     }
 
-    void checkConfinementIfNeeded() {
-        Thread ownerThread = scope().ownerThread();
-        if (ownerThread != null && ownerThread != Thread.currentThread()) {
-            throw new IllegalStateException("Attempt to allocate outside confinement thread");
-        }
+    public ResourceScope scope() {
+        return scope;
     }
 
-    ResourceScope scope() {
-        return segment.scope();
+    private MemorySegment newSegment(long size, long align) {
+        return MemorySegment.allocateNative(size, align, scope);
     }
 
-    public static class UnboundedArenaAllocator extends ArenaAllocator {
-
-        private static final long DEFAULT_BLOCK_SIZE = 4 * 1024;
-
-        public UnboundedArenaAllocator(ResourceScope scope) {
-            super(MemorySegment.allocateNative(DEFAULT_BLOCK_SIZE, 1, scope));
-        }
-
-        private MemorySegment newSegment(long size, long align) {
-            return MemorySegment.allocateNative(size, align, segment.scope());
-        }
-
-        @Override
-        public MemorySegment allocate(long bytesSize, long bytesAlignment) {
-            checkConfinementIfNeeded();
+    @Override
+    public MemorySegment allocate(long bytesSize, long bytesAlignment) {
+        long prevSp = sp;
+        long allocatedSize = 0L;
+        try {
             // try to slice from current segment first...
             MemorySegment slice = trySlice(bytesSize, bytesAlignment);
             if (slice != null) {
+                allocatedSize = sp - prevSp;
                 return slice;
             } else {
                 long maxPossibleAllocationSize = bytesSize + bytesAlignment - 1;
-                if (maxPossibleAllocationSize > DEFAULT_BLOCK_SIZE) {
+                if (maxPossibleAllocationSize > blockSize) {
                     // too big
+                    allocatedSize = Utils.alignUp(bytesSize, bytesAlignment);
+                    if (size > arenaSize) {
+                        throw new OutOfMemoryError();
+                    }
```

Comment:
I'll have to think some more about this. I don't think this is covered inside the block - that is, the block tries to allocate, and then in the finally we throw if we realized we've allocated too much.

Reply:
What is missing, I think, is a check (size > arenaSize) at the beginning of the method (we only check this in one of the paths). But we need to check before and after, I think, as it is possible to allocate a segment and then realize that we ended up overflowing the arena size.

Reply:
While what I said above correctly reflects what the implementation does, I think a broader issue is that the arena allocator implementation is allocating sometimes more native memory than what its contract specifies. While in some cases we can prevent that, I think in the general case (e.g. where we allocate a new block) we cannot, unless we add extra API guarantees - e.g. that the arena size should be a multiple of the block size (but then we'd have to special case `Long.MAX_VALUE`, or maybe pick a "big enough" power of two instead)

Reply:
Maybe we should not support block size in the case of a bounded arena. i.e. just allocate the whole thing upfront, and have 3 APIs:

1. arena with no bounds and default block size.
2. arena with no bounds and custom block size.
3. arena with bounds, that has no blocks size but allocates the whole thing in one go (could be modeled as block size = arena size).

Right now we have 1. and 2., but instead of 3. we have a variant that allows setting both the arena size and block size.

If we want to keep what we currently have, I'd suggest changing the arena size to a block count for the variant that takes both the arena size and the block size.

Any ways, that seems like something we should address for 19 maybe, not now.